### PR TITLE
llms: Add mistral hosted inference llm implementation

### LINKED ISSUE #486: llms: Implement Mistral Platform
We should add [Mistral Platform](https://docs.mistral.ai/) support. They point to a pre-existing Go [client](https://docs.mistral.ai/platform/client/) at https://github.com/Gage-Technologies/mistral-go but I haven't evaluated it yet.

### PR DESCRIPTION
**Pull request by aannirajpatel**
Adds an LLM implementation to leverage the [Mistral Platform](https://mistral.ai). Wraps [mistral-go](https://github.com/Gage-Technologies/mistral-go).

Requesting feedback on Function calling-related implementation.

Checklist
✅ I have read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
✅ I have read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
✅ I checked that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
✅ My PR provides a description that addresses what it is solving, with a focus on integrating Mistral LLMs into LangChain.
✅ My PR describes the source of new concepts introduced (no new concepts introduced).
✅ My PR references existing implementations as appropriate, primarily the mistral-go client library documentation.
🟡 PR should contain test coverage for new functions (to be added - **question to maintainers**: can these be added retrospectively?).
✅ Verified that my PR passes all [golangci-lint](https://golangci-lint.run/) checks.

resolves #486

### CODE REVIEW

File: llms/mistral/mistralmodel.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	if len(res.Choices) != 1 {
+		m.CallbacksHandler.HandleLLMError(ctx, err)
+		return "", errors.New("unexpected response from Mistral SDK, length of the Choices slice must be 1")
+	}
+
+	return res.Choices[0].Message.Content, nil
+}
+
+// Supported models: https://docs.mistral.ai/platform/endpoints/
+// TODO: Mistral also supports ResponseType, which, when set to "json", ensures the model's output is strictly a JSON object. Should this be made a part of client config and pulled whenever Call or GenerateContent is called?
```
Comment (tmc) [inline llms/mistral/mistralmodel.go:45]: this comment isn't well formed enough to accept this PR,  if you want to leave this in the code please remove it from the docstring of this (albeit internal) function.

Comment (aannirajpatel) [reply llms/mistral/mistralmodel.go]: Got it; go longer a docstring.

Comment (tmc) [reply llms/mistral/mistralmodel.go]: Not a big deal just want to make sure that the contents guide the next contributor effectively!

Comment (aannirajpatel) [reply llms/mistral/mistralmodel.go]: Makes sense, thanks for the prompt review!

File: llms/mistral/client_options.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	apiKey     string
+	endpoint   string
+	maxRetries int
+	timeout    time.Duration
+	model      string
+}
+
+type Option func(*clientOptions)
+
+func WithAPIKey(apiKey string) Option {
```
Comment (tmc) [inline llms/mistral/client_options.go:15]: please include comments on exported symbols

File: llms/mistral/mistralmodel.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+	"errors"
+	"os"
+
+	sdk "github.com/gage-technologies/mistral-go"
+	"github.com/tmc/langchaingo/callbacks"
+	"github.com/tmc/langchaingo/llms"
+	"github.com/tmc/langchaingo/schema"
+)
+
+// LLM is a mistral LLM implementation.
```
Comment (tmc) [inline llms/mistral/mistralmodel.go:14]: s/a m/a M/

File: llms/mistral/mistralmodel.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+type LLM struct {
+	CallbacksHandler callbacks.Handler
```
Comment (tmc) [inline llms/mistral/mistralmodel.go:16]: not necessary but given how many bytes/tokens will flow through here I feel like some comments would be fitting.

**Review by tmc — COMMENTED** [review]
If only this were before the Mistral hack in SF this weekend! With a few touches I think this is going to be good to go in!

**Comment by aannirajpatel** [comment]
> If only this were before the Mistral hack in SF this weekend! With a few touches I think this is going to be good to go in!

Ahh, missed it 😅 Just learned about that on social media. Cool hackathon!

File: examples/mistral-completion-example/mistral_completion_example.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/tmc/langchaingo/llms"
+	"github.com/tmc/langchaingo/llms/mistral"
+)
+
+func main() {
+	llm, err := mistral.New(mistral.WithAPIKey("API_KEY_GOES_HERE"))
```
Comment (tmc) [inline examples/mistral-completion-example/mistral_completion_example.go:13]: not blocking but I can't help but think that this comment could be laid out in a manner that would be more accessible -- could you add a comment either at the program/file level in a way that would help new folks figure out what's going on? Even better would be a docs-site addition along a blog post.

File: examples/mistral-completion-example/mistral_completion_example.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		llms.WithTemperature(0.8),
+		llms.WithStreamingFunc(func(ctx context.Context, chunk []byte) error {
+			fmt.Print(string(chunk))
+			return nil
+		}),
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	_ = completionWithStreaming
```
Comment (tmc) [inline examples/mistral-completion-example/mistral_completion_example.go:35]: folks that know go well may understand why this line is here but a comment would be useful

File: examples/mistral-completion-example/mistral_completion_example.go (hunk 5 above)
Comment (aannirajpatel) [reply examples/mistral-completion-example/mistral_completion_example.go:13]: Yes, that doc site is cool! Will add some docs over this week.

Comment (aannirajpatel) [reply examples/mistral-completion-example/mistral_completion_example.go]: Added some docs for now.

File: llms/mistral/client_options.go (hunk 2 above)
Comment (aannirajpatel) [reply llms/mistral/client_options.go:21]: Will fix 👍

**Comment by tmc** [comment]
Really impressed with the docs additon! Thanks so much, I'd like to normalize that new contributors and would love any input you have to make that more accessible or straightforward.

I commented with a note about how we tend to start comments on exported functions/types in this codebase, I think that's the only snag that remains here!

**Comment by tmc** [comment]
🙏 merged.

File: llms/mistral/client_options.go (hunk 2 above)
Comment (aannirajpatel) [reply llms/mistral/client_options.go:21]: I see it's already merged. I'll change these up as a follow-up PR on the coming weekend.

**Pull request merged**